pkg/kademlia/queue.go

```diff
@@ -4,57 +4,108 @@
 package kademlia
 
 import (
+	"container/heap"
 	"math/big"
 
+	"storj.io/storj/pkg/dht"
 	"storj.io/storj/pkg/pb"
 )
 
-// An Item is something we manage in a priority queue.
-type Item struct {
+//We've encapsulated our priorityQueue{} implementation (below) so that you don't
+//have to know how the Golang heap API works (nor our priority XOR logic)
+
+//XorQueue is a priority queue where the priority is key XOR distance
+type XorQueue struct {
+	pq *priorityQueue
+}
+
+//NewXorQueue returns a priorityQueue with priority based on XOR from targetBytes
+func NewXorQueue(nodes []*pb.Node, target dht.NodeID) XorQueue {
+	targetBytes := new(big.Int).SetBytes(target.Bytes())
+	pq := make(priorityQueue, len(nodes))
+	for i, node := range nodes {
+		pq[i] = &item{value: node, index: i}
+		pq[i].priority = new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId())))
+	}
+	heap.Init(&pq)
+	return XorQueue{pq: &pq} 
+}
+
+//Insert adds Node onto the queue
+func (x XorQueue) Insert(node *pb.Node, target dht.NodeID) {
+	targetBytes := new(big.Int).SetBytes(target.Bytes())
+	heap.Push(x.pq, &item{
+		value:    node,
+		priority: new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId()))),
+	})
+}
```

Comment:
Handle the queue limit internally rather than pushing that responsibility to the user.

Reply:
agreed